File: src/persist.js
// ============================================================
// persist.js — One-function database persistence for y8 Prolog
//
// Uses engine.onAssert / engine.onRetract callbacks — no
// monkey-patching.  Ephemeral scopes become SQL transactions.
//
// Usage:
//   persist(engine, sqliteAdapter(db));
//   persist(engine, db);  // auto-detect better-sqlite3
//   persist(engine, adapter, null, {stringify: qjson_stringify, parse: qjson_parse});
//
// Adapter interface (6 methods):
//   setup, insert(key,functor,arity), remove(key), all(predicates), commit, close
// ============================================================

// ── Term serialization (inline, matches sync.js format) ─────

function _ser(t) {
  if (t.type === "atom") return { t: "a", n: t.name };
  if (t.type === "num") {
    var o = { t: "n", v: t.value };
    if (t.repr) o.r = t.repr;
    return o;
  }
  if (t.type === "compound") {
    var a = [];
    for (var i = 0; i < t.args.length; i++) a.push(_ser(t.args[i]));
    return { t: "c", f: t.functor, a: a };
  }
  return null;
}

function _deser(o) {
  if (o.t === "a") return { type: "atom", name: o.n };
  if (o.t === "n") {
    var t = { type: "num", value: o.v };
    if (o.r) t.repr = o.r;
    return t;
  }
  if (o.t === "c") {
    var a = [];
    for (var i = 0; i < o.a.length; i++) a.push(_deser(o.a[i]));
    return { type: "compound", functor: o.f, args: a };
  }
  return null;
}

// ── Auto-detect better-sqlite3 → semantic adapter ───────────

function _autoAdapter(db) {
  if (typeof db.insert === "function" && typeof db.setup === "function") {
    return db;
  }
  if (typeof db.prepare === "function" && typeof db.exec === "function") {
    var cache = {};
    function stmt(sql) {
      if (!cache[sql]) cache[sql] = db.prepare(sql);
      return cache[sql];
    }
    return {
      setup:  function() {
        db.exec("CREATE TABLE IF NOT EXISTS facts (term TEXT PRIMARY KEY, functor TEXT, arity INTEGER)");
        db.exec("CREATE INDEX IF NOT EXISTS idx_facts_pred ON facts(functor, arity)");
      },
      insert: function(key, functor, arity) { stmt("INSERT OR IGNORE INTO facts VALUES (?, ?, ?)").run(key, functor, arity); },
      remove: function(key) { stmt("DELETE FROM facts WHERE term = ?").run(key); },
      all:    function(predicates) {
        if (predicates) {
          var rows = [], keys = Object.keys(predicates);
          for (var i = 0; i < keys.length; i++) {
            var parts = keys[i].split("/");
            var matched = stmt("SELECT term FROM facts WHERE functor = ? AND arity = ?").all(parts[0], parseInt(parts[1], 10));
            for (var j = 0; j < matched.length; j++) rows.push(matched[j].term);
          }
          return rows;
        }
        return stmt("SELECT term FROM facts").all().map(function(r) { return r.term; });
      },
      commit: function() {},
      close:  function() { db.close(); }
    };
  }
  return db;
}

// ── Main function ───────────────────────────────────────────

function persist(engine, db, predicates, codec) {
  var adapter = _autoAdapter(db);

  var _dumps = (codec && codec.stringify) || JSON.stringify;
  var _codec_parse = codec && codec.parse;
  var _loads = _codec_parse
    ? function(text) { try { return JSON.parse(text); } catch(e) { return _codec_parse(text); } }
    : JSON.parse;

  function _key(term) { return _dumps(_ser(term)); }

  function _pred(term) {
    if (term.type === "compound") return [term.functor, term.args.length];
    if (term.type === "atom") return [term.name, 0];
    return [null, null];
  }

  var preds = predicates || null;
  var txnDepth = 0;

  function _ok(term) {
    if (!preds) return true;
    var key;
    if (term.type === "compound") key = term.functor + "/" + term.args.length;
    else if (term.type === "atom") key = term.name + "/0";
    else return false;
    return !!preds[key];
  }

  function _commit() {
    if (txnDepth === 0 && adapter.commit) adapter.commit();
  }

  // ── Restore ───────────────────────────────────────────
  adapter.setup();
  var keys = adapter.all(preds);
  for (var i = 0; i < keys.length; i++) {
    engine.addClause(_deser(_loads(keys[i])));
  }

  // ── Listen for assert → INSERT ────────────────────────
  engine.onAssert.push(function(head) {
    if (_ok(head)) {
      var p = _pred(head);
      adapter.insert(_key(head), p[0], p[1]);
      _commit();
    }
  });

  // ── Listen for retract → DELETE ───────────────────────
  engine.onRetract.push(function(head) {
    if (_ok(head)) {
      adapter.remove(_key(head));
      _commit();
    }
  });

  // ── Hook ephemeral/1 — transaction batching ───────────
  if (engine.builtins["ephemeral/1"]) {
    var origEphemeral = engine.builtins["ephemeral/1"];
    engine.builtins["ephemeral/1"] = function(goal, rest, subst, counter, depth, onSolution) {
      txnDepth++;
      try {
        origEphemeral(goal, rest, subst, counter, depth, onSolution);
      } finally {
        txnDepth--;
        if (txnDepth === 0 && adapter.commit) adapter.commit();
      }
    };
  }

  return adapter;
}

// ── Export (dual ESM/CJS) ───────────────────────────────────

if (typeof exports !== "undefined") {
  exports.persist = persist;
}
export { persist };

File: src/persist.py
# ============================================================
# persist.py — One-function database persistence for Y@ Prolog
#
# Usage:
#   from prolog import Engine
#   from persist import persist
#
#   engine = Engine()
#   persist(engine, "state.db")           # SQLite shorthand
#   persist(engine, sqlite_adapter(path)) # explicit adapter
#   persist(engine, adapter, codec="qjson") # BigNum support
#
# Uses engine.on_assert / engine.on_retract callbacks — no
# monkey-patching.  Ephemeral scopes become SQL transactions.
# ============================================================

import json
from prolog import deep_walk, unify


def _ser(term):
    if term[0] == "atom":     return {"t": "a", "n": term[1]}
    if term[0] == "num":
        o = {"t": "n", "v": term[1]}
        if len(term) > 2 and term[2]:
            o["r"] = term[2]
        return o
    if term[0] == "compound":
        return {"t": "c", "f": term[1], "a": [_ser(a) for a in term[2]]}
    return None


def _deser(obj):
    if obj["t"] == "a": return ("atom", obj["n"])
    if obj["t"] == "n":
        if "r" in obj:
            return ("num", obj["v"], obj["r"])
        return ("num", obj["v"])
    if obj["t"] == "c":
        return ("compound", obj["f"], tuple(_deser(a) for a in obj["a"]))
    return None


def _resolve_adapter(db):
    if isinstance(db, str):
        from persist_sqlite import sqlite_adapter
        return sqlite_adapter(db)
    if isinstance(db, dict) and "insert" in db:
        return db
    from persist_pg import pg_adapter
    return pg_adapter(db)


def _resolve_codec(codec):
    if codec == "qjson":
        from qjson import stringify as _qs, parse as _qp
        def _loads(text):
            try:
                return json.loads(text)
            except (ValueError, TypeError):
                return _qp(text)
        return _qs, _loads
    if codec and not callable(codec):
        _cp = codec.get("parse")
        _cs = codec.get("stringify", lambda o: json.dumps(o, separators=(',', ':')))
        if _cp:
            def _loads(text):
                try:
                    return json.loads(text)
                except (ValueError, TypeError):
                    return _cp(text)
            return _cs, _loads
        return _cs, json.loads
    return lambda obj: json.dumps(obj, separators=(',', ':')), json.loads


def persist(engine, db, predicates=None, codec=None):
    """Attach database persistence to a Prolog engine.

    db         — file path (SQLite), adapter dict, or DBAPI 2.0 connection
    predicates — set of "functor/arity" to persist; None = all dynamic facts
    codec      — "qjson" for BigInt/BigDecimal/BigFloat, or None for plain JSON

    Returns the adapter.
    """
    adapter = _resolve_adapter(db)
    _dumps, _loads = _resolve_codec(codec)

    def _key(term):
        return _dumps(_ser(term))

    def _pred(term):
        if term[0] == "compound":
            return term[1], len(term[2])
        if term[0] == "atom":
            return term[1], 0
        return None, None

    def _ok(term):
        if predicates is None:
            return True
        if term[0] == "compound":
            return term[1] + "/" + str(len(term[2])) in predicates
        if term[0] == "atom":
            return term[1] + "/0" in predicates
        return False

    _txn = [0]

    def _commit():
        if _txn[0] == 0:
            adapter["commit"]()

    # ── Restore saved facts ──────────────────────────────────
    adapter["setup"]()
    for text in adapter["all"](predicates):
        engine.add_clause(_deser(_loads(text)))

    # ── Listen for assert → INSERT ───────────────────────────
    def _on_assert(head):
        if _ok(head):
            f, a = _pred(head)
            adapter["insert"](_key(head), f, a)
            _commit()

    engine.on_assert.append(_on_assert)

    # ── Listen for retract → DELETE ──────────────────────────
    def _on_retract(head):
        if _ok(head):
            adapter["remove"](_key(head))
            _commit()

    engine.on_retract.append(_on_retract)

    # ── Hook ephemeral/1 — ephemeral scope = SQL transaction ─
    if "ephemeral/1" in engine.builtins:
        _orig_eph = engine.builtins["ephemeral/1"]

        def _hooked_ephemeral(goal, rest, subst, depth, on_sol):
            _txn[0] += 1
            try:
                _orig_eph(goal, rest, subst, depth, on_sol)
            finally:
                _txn[0] -= 1
                if _txn[0] == 0:
                    adapter["commit"]()

        engine.builtins["ephemeral/1"] = _hooked_ephemeral

    return adapter
